From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Thu, 24 Jul 2025 14:07:47 +0300
Subject: [PATCH] Optimize collections


diff --git a/net/minecraft/core/NonNullList.java b/net/minecraft/core/NonNullList.java
index 14bd5efe0da5c268f9b8fc014c4a14b68dcbee8d..352856ade8cbbbbeb90961de3b65d9a94a5ca360 100644
--- a/net/minecraft/core/NonNullList.java
+++ b/net/minecraft/core/NonNullList.java
@@ -12,23 +12,23 @@ public class NonNullList<E> extends AbstractList<E> {
     private final @Nullable E defaultValue;
 
     public static <E> NonNullList<E> create() {
-        return new NonNullList<>(Lists.newArrayList(), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(), null); // DivineMC - Optimize collections
     }
 
     public static <E> NonNullList<E> createWithCapacity(final int capacity) {
-        return new NonNullList<>(Lists.newArrayListWithCapacity(capacity), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(capacity), null); // DivineMC - Optimize collections
     }
 
     public static <E> NonNullList<E> withSize(final int size, final E defaultValue) {
         Objects.requireNonNull(defaultValue);
         Object[] objects = new Object[size];
         Arrays.fill(objects, defaultValue);
-        return new NonNullList<>(Arrays.asList((E[])objects), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>((E[])objects), defaultValue); // DivineMC - Optimize collections
     }
 
     @SafeVarargs
     public static <E> NonNullList<E> of(final E defaultValue, final E... values) {
-        return new NonNullList<>(Arrays.asList(values), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(values), defaultValue); // DivineMC - Optimize collections
     }
 
     protected NonNullList(final List<E> list, final @Nullable E defaultValue) {
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 7d7962fbf2f9d6068e43978ddc7f20d0bc12d41a..179627ccce48610c9bff3da9d4aaf51182002b90 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -139,8 +139,10 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
     public final ChunkMap.DistanceManager distanceManager;
     private final String storageName;
     private final PlayerMap playerMap = new PlayerMap();
-    public final Int2ObjectMap<ChunkMap.TrackedEntity> entityMap = new Int2ObjectOpenHashMap<>();
-    private final Long2ByteMap chunkTypeCache = new Long2ByteOpenHashMap();
+    // DivineMC start - Optimize collections
+    public final Int2ObjectMap<ChunkMap.TrackedEntity> entityMap = new org.bxteam.divinemc.util.map.Int2ObjectConcurrentHashMap<>();
+    private final Long2ByteMap chunkTypeCache = it.unimi.dsi.fastutil.longs.Long2ByteMaps.synchronize(new Long2ByteOpenHashMap());
+    // DivineMC end - Optimize collections
     // Paper - rewrite chunk system
     public int serverViewDistance;
     public final WorldGenContext worldGenContext; // Paper - public
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 3fa2712d0b1b1361fb4e9c6bc49440f0c092144c..ea46127eca64c17c414e373948473691d2a713e5 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -267,7 +267,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     @Override
     public final <T extends Entity> List<T> getEntitiesOfClass(final Class<T> entityClass, final AABB boundingBox, final Predicate<? super T> predicate) {
         Profiler.get().incrementCounter("getEntities");
-        final List<T> ret = new java.util.ArrayList<>();
+        final List<T> ret = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(); // DivineMC - Optimize collections
 
         ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)this).moonrise$getEntityLookup().getEntities(entityClass, null, boundingBox, ret, predicate);
 
@@ -277,7 +277,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     @Override
     public final List<Entity> moonrise$getHardCollidingEntities(final Entity entity, final AABB box, final Predicate<? super Entity> predicate) {
         Profiler.get().incrementCounter("getEntities");
-        final List<Entity> ret = new java.util.ArrayList<>();
+        final List<Entity> ret = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(); // DivineMC - Optimize collections
 
         ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)this).moonrise$getEntityLookup().getHardCollidingEntities(entity, box, ret, predicate);
 
diff --git a/net/minecraft/world/level/chunk/ProtoChunk.java b/net/minecraft/world/level/chunk/ProtoChunk.java
index d3bde6eb0edd24d5e5352388788d30938712a9c7..d4ace19027214e5ed83c83a2cd97f1bb098d8152 100644
--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -44,7 +44,7 @@ public class ProtoChunk extends ChunkAccess {
     private static final Logger LOGGER = LogUtils.getLogger();
     private volatile @Nullable LevelLightEngine lightEngine;
     private volatile ChunkStatus status = ChunkStatus.EMPTY;
-    private final List<CompoundTag> entities = Lists.newArrayList();
+    private final List<CompoundTag> entities = Collections.synchronizedList(Lists.newArrayList()); // DivineMC - Optimize collections
     private @Nullable CarvingMask carvingMask;
     private @Nullable BelowZeroRetrogen belowZeroRetrogen;
     private final ProtoChunkTicks<Block> blockTicks;
